keps/sig-cli/2590-kubectl-subresource/README.md

```diff
@@ -0,0 +1,393 @@
+
+
+### Non-Goals
+
+- Support subresources other than `status` and `scale`.
+- Allow updates to the status subresource if a controller is running against the main
+  resource that reconciles `.status` part of the resource.
+- Allow specifying `additionalPrinterColumns` for CRDs for the status subresource.
+
+## Proposal
+
+kubectl commands like get, apply, patch, edit and replace will now contain a
+new flag `--subresource=[subresource-name]` which will allow fetching and updating
+`status` and `scale` subresources for all API resources.
+
+When the `--subresource=status` flag is used against built-in API types, 
+the `.status` field of the resource is _not_ updated since the built-in controllers run against
+the resource and reconcile the `.status` part. This also holds true for custom resources if a 
+custom controller reconciles the `.status` part of the CR. It ensures that `.status` of 
+built-in API types will never store arbitrary data.
+
+If `--subresource` flag is used for a resource that doesn't support the subresource, 
+a `NotFound` error will be returned.
+
+### Examples
+
+#### get
+
+```shell
+# for built-in types
+$ kubectl get deployment nginx-deployment --subresource=status
+NAME               READY   UP-TO-DATE   AVAILABLE   AGE
+nginx-deployment   3/3     3            3           43s
+
+$ kubectl get deployment nginx-deployment --subresource=scale
+NAME               DESIREDREPLICAS   AVAILABLEREPLICAS
+nginx-deployment   3                 3
+
+# for CRDS
+$ kubectl get crontab cron--subresource=status
+NAME   SPEC          REPLICAS   AGE
+cron   * * * * */5   3          4m52s
+
+$ kubectl get vmset vmset-1 --subresource=scale
+NAME   DESIREDREPLICAS   AVAILABLEREPLICAS
+cron   3                 0
+```
+
+Invalid subresources:
+
+```shell
+$ kubectl get pod nginx-deployment-66b6c48dd5-dv6gl --subresource=logs
+error: --subresource must be one of [status scale], not "logs"
+
+$ kubectl get pod nginx-deployment-66b6c48dd5-dv6gl --subresource=scale
+Error from server (NotFound): the server could not find the requested resource
+```
+
+#### patch
+
+```shell
+# For built-in types
+# update spec.replicas through scale subresource
+$ kubectl patch deployment nginx-deployment --subresource='scale' --type='merge' -p '{"spec":{"replicas":2}}'
+scale.autoscaling/nginx-deployment patched
+
+# spec.replicas is updated for the main resource
+$ k get deploy nginx-deployment
+NAME               READY   UP-TO-DATE   AVAILABLE   AGE
+nginx-deployment   2/2     2            2           4m
+
+# For CRDs
+$ kubectl patch crontabs cron --subresource='status' --type='merge' -p '{"status":{"replicas":2}}'
+crontab.stable.example.com/cron patched
+
+$ kubectl get crontab cron --subresource=scale
+NAME   DESIREDREPLICAS   AVAILABLEREPLICAS
+cron   3                 2
+```
+
+
+#### apply
+
+```shell
+# cron-with-status.yaml has .status.replicas=3
+$ kubectl apply -f cron-with-status.yaml --subresource=status
```

Comment:
This will override the `last-applied` annotation set by another `kubectl apply` occurence (probably on main resource), which will result in odd behaviors later. Also note that some resources don't allow annotations to be updated through status subresource. That would probably work with server-side apply.

Reply:
This potentially odd behavior was brought up during the SIG-CLI demo too. I am currently investigating what should be done to handle this properly.

Reply:
There was another question I brought during the call,  if the argument contains just status or full object?

Reply:
when using `apply` with subresource it turns out the `last-applied` annotation is not modified. The annotation is under the metadata section and that section and any changes to the updates are ignored when modifying the subresources.

Therefore any subsequent `apply` calls to the resource itself are not affected as the patch calculated for the `apply` call will use the correct `last-applied` configuration of _resource_. 

The same cannot be said when `apply`-ing a subresource. The patch calculated using the annotation could be wrong because the last applied configuration of the status is not captured in the annotation.